### DISCUSSION TITLE
Invalid Body ID exceptions

### DISCUSSION THREAD

**Discussion by JasperKr** [jrouwe/JoltPhysics#1191] [Q&A]
I keep getting an exception on this line:
`JPH_ASSERT((inID & cBroadPhaseBit) == 0 || inID == cInvalidBodyID); // Check bit used by broadphase`
it's line 32 in BodyID.h

I never edit the body id's so i'm not sure what's causing this to happen.

It sometimes happens when i create a new physics object, then resize it. 
My resizing code deletes the physics object and immediately create a new one,
i do this so i can scale objects in the editor in runtime, i have to do it immediately since i use jolt's raycasts for object selection.

Unless someone has suggestions on how to do it i'm not quite sure how to attach a debugger. 
I run jolt using a FFI and the C-binding from JoltPhysicsSharp, so i'm not sure if can check the actual value of the id when it crashes.

**Comment by jrouwe** [comment]
Hello,

Could you please post a callstack for when the assert triggers and the value of inID? 

I have no idea why this assert would trigger unless you specified a max number of bodies > BodyID::cMaxBodyIndex (0x7fffff) in PhysicsSystem::Init and actually have created this many bodies.

Note that to dynamically scale a body, you don't need to create a new body. https://jrouwe.github.io/JoltPhysics/#scaling-shapes explains how to do this and includes a demo.

**Reply by JasperKr** [reply]
Hi, i don't set the max number of bodies. When loading the engine there are 0 other bodies so there shouldn't be an issue with that i don't think.
About the callstack, i'm still not exactly sure how to get better info, but, i've ran it with gdb until i encountered the exception, since it doesn't always seem to happen.
```
C:\Coding\Forks\JoltPhysicsSharp\build\vs2022_x64_double\_deps\joltphysics-src\Jolt\Physics\Body\BodyID.h:32: ((inID & cBroadPhaseBit) == 0 || inID == cInvalidBodyID)

Thread 1 received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffecfd371b4 in joltc_double!JPH_HingeConstraint_SetMotorState ()
   from C:\Coding\LOVE\Rhodium\V3\Engine\bin\lib\joltc_double3.dll
(gdb) backtrace
#0  0x00007ffecfd371b4 in joltc_double!JPH_HingeConstraint_SetMotorState ()
   from C:\Coding\LOVE\Rhodium\V3\Engine\bin\lib\joltc_double3.dll
#1  0x00007ffecfd21523 in joltc_double!JPH_HingeConstraint_SetMotorState ()
   from C:\Coding\LOVE\Rhodium\V3\Engine\bin\lib\joltc_double3.dll
#2  0x00007fff69f35110 in ?? () from C:\Coding\LOVE\Rhodium\V3\Engine\lua51.dll
#3  0x00007fff69f7b5c9 in lua51!fwscanf_s () from C:\Coding\LOVE\Rhodium\V3\Engine\lua51.dll
#4  0x00007fff69f49632 in ?? () from C:\Coding\LOVE\Rhodium\V3\Engine\lua51.dll
#5  0x00007fff69f32cb0 in ?? () from C:\Coding\LOVE\Rhodium\V3\Engine\lua51.dll
#6  0x00007ff77c2b13b5 in lovec!?getType@Variant@love@@QEBA?AW4Type@12@XZ ()
#7  0x00007ff77c2b1544 in lovec!?unknown@Variant@love@@SA?AV12@XZ ()
#8  0x00007ff77c2b2570 in lovec!?unknown@Variant@love@@SA?AV12@XZ ()
#9  0x00007ff77c2b19e0 in lovec!?unknown@Variant@love@@SA?AV12@XZ ()
#10 0x00007fffb19d7374 in KERNEL32!BaseThreadInitThunk () from C:\Windows\System32\kernel32.dll
#11 0x00007fffb2afcc91 in ntdll!RtlUserThreadStart () from C:\Windows\SYSTEM32\ntdll.dll
#12 0x0000000000000000 in ?? ()
```
(for context, LÖVE is the game framework i use)

There is some stuff about a hinge constraint in there which doesn't make much sense to me, since i never create one.

About the scaled shapes, it looks like the C api doesn't expose any of that at the moment, otherwise i'd switch over to using those.

Hope this helps.

**Reply by jrouwe** [reply]
This doesn't make sense to me either. The JPH_HingeConstraint_SetMotorState function also doesn't call that constructor in BodyID.h, so it sounds to me like there is some sort of problem in the integration. 

Browsing the code of LOVE, I don't think they have integrated Jolt, so did you do this yourself? How do you bind the C api into LOVE and is there perhaps a mismatch in the compiler defines between joltc_double3.dll and the place where it is being called from? (e.g. the define JPH_DOUBLE_PRECISION, JPH_DEBUG_RENDERER, JPH_OBJECT_LAYER_BITS and JPH_OBJECT_STREAM and possibly others mentioned in RegisterTypesInternal have an effect on the structure sizes and could potentially lead to strange memory corruptions).

It is really hard for me to say anything sensible about this as it is all code that is external to Jolt.

**Reply by JasperKr** [reply]
Yeah i've integrated it using luaJIT's C FFI, which interfaces with that C binding from JoltPhysicsSharp, i load the dll and define the functions as provided in this file:
https://github.com/amerkoleci/JoltPhysicsSharp/blob/main/src/joltc/joltc.h,
I don't define extra stuff like `JPH_DOUBLE_PRECISION`, however the build.bat i use to build it does have the `-DDOUBLE_PRECISION=ON` argument, should i define it elsewhere too?
Otherwise i'll probably just try compiling for the non-double precision version and see if that fixes it.

**Reply by jrouwe** [reply]
joltc.h uses the JPH_DOUBLE_PRECISION define, so make sure that it matches what the dll was created with when importing it.

**Reply by JasperKr** [reply]
I still get the same issue without double precision so it must have something to do with the binding / the way i interface with it. Thanks the help though!
